### PR TITLE
Rewrote C code with C++ extensions

### PR DESCRIPTION
**Pull request by goldsborough**
Rewrote the C backend with our new C++ extensions. Much less config, much less dependencies (-cffi), much less code, much more happy.

Test plan:
- Ran existing tests,
- Wrote additional test to make sure that loading -> writing -> loading is an identity operation,
- Loaded an mp3 and wav file, and re-saved them. Listened to them and verified subjectively they're the same.

@soumith @colesbury

### CODE REVIEW

**Comment by goldsborough** [comment]
Will add more tests. Still WIP.

File: torchaudio/torch_sox.cpp [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        "Error reading audio file: empty file or read failed in sox_read");
+  }
+
+  output.resize_({samples_read / number_of_channels, number_of_channels});
+  output = output.contiguous();
+
+  AT_DISPATCH_ALL_TYPES(output.type(), "read_audio_buffer", [&] {
+    auto* data = output.data<scalar_t>();
+    std::copy(buffer.begin(), buffer.begin() + samples_read, data);
+  });
```
Comment (ezyang) [inline torchaudio/torch_sox.cpp]: It doesn't look like you ever `sox_close(fd)` here

File: torchaudio/torch_sox.cpp [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    const std::string& extension,
+    int sample_rate) {
+  if (sox_init() != SOX_SUCCESS) {
+    throw std::runtime_error("Error initializing sox library");
+  }
+
+  if (!tensor.is_contiguous()) {
+    throw std::runtime_error(
+        "Error writing audio file: input tensor must be contiguous");
+  }
```
Comment (ezyang) [inline torchaudio/torch_sox.cpp]: This is missing a test that the input tensor is 2D. (The original code was missing this too, but it didn't use `numel` to compute length, so you wouldn't ever read out of bounds. This code can easily read out of bounds if you have a multi-dim tensor.)

File: torchaudio/torch_sox.cpp [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+      /*encoding=*/nullptr,
+      extension.c_str(),
+      /*filetype=*/nullptr,
+      /*oob=*/nullptr);
+
+  if (fd == nullptr) {
+    throw std::runtime_error(
+        "Error writing audio file: could not open file for writing");
+  }
+
```
Comment (ezyang) [inline torchaudio/torch_sox.cpp]: I'm not sure what the merit of deleting the `read_audio` and `write_audio` helper functions was; they seem like a reasonable code organization.

File: torchaudio/torch_sox.cpp [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+  std::vector<sox_sample_t> buffer(tensor.numel());
```
Comment (ezyang) [inline torchaudio/torch_sox.cpp]: This code is a bit perplexing. `sox_sample_t` is a 32-bit integer. So why don't you just convert the tensor into an `IntTensor` and then write samples out straight from the internal tensor buffer?

File: torchaudio/torch_sox.cpp [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+  AT_DISPATCH_ALL_TYPES(tensor.type(), "write_audio_buffer", [&] {
+    auto* data = tensor.data<scalar_t>();
+    std::copy(data, data + tensor.numel(), buffer.begin());
+  });
+
+  const auto samples_written = sox_write(fd, buffer.data(), buffer.size());
+
+  // Free buffer and sox structures (before possibly throwing).
+  sox_close(fd);
+  sox_quit();
```
Comment (ezyang) [inline torchaudio/torch_sox.cpp]: Another quit here

File: torchaudio/torch_sox.cpp (hunk 4 above)
Comment (goldsborough) [reply torchaudio/torch_sox.cpp]: We sometimes pass `FloatTensor` to this code, so we need to convert to integers first

**Pull request merged**